mini-projects/01-device-driver/laptop-interface/src/gui.rs

```diff
@@ -0,0 +1,390 @@
+            .draw_series(LineSeries::new(data2, &BLUE))
+            .unwrap()
+            .label(label2)
+            .legend(|(x, y)| Circle::new((x, y), 3, BLUE));
+
+        chart.configure_series_labels().draw().unwrap();
+        root.present().unwrap();
+    }
+
+    // Function to compute displacement from positions (time, (x, y))
+    fn compute_displacement(&self, positions: &[(f64, (i32, i32))]) -> Vec<(f64, f64)> {
+        positions
+            .iter()
+            .map(|(t, (x, y))| (*t, ((x.pow(2) + y.pow(2)) as f64).sqrt()))
+            .collect()
+    }
+
+    // Plot positions for controllers
+    fn plot_positions(&self) {
+        let positions_one = self.compute_displacement(&self.controller_one_pos);
+        let positions_two = self.compute_displacement(&self.controller_two_pos);
+        self.plot_data(
+            "positions.png",
+            "Position vs Time",
+            "Controller 0",
+            "Controller 1",
+            positions_one,
+            positions_two,
+        );
+    }
+
+    // Plot displacement differences between controllers
+    fn plot_difference(&self) {
+        let positions_one = self.compute_displacement(&self.controller_one_pos);
+        let positions_two = self.compute_displacement(&self.controller_two_pos);
+
+        let difference_one: Vec<(f64, f64)> = positions_one
+            .iter()
+            .zip(positions_two.iter())
+            .map(|((t1, d1), (_, d2))| (*t1, d1 - d2))
+            .collect();
+
+        let difference_two: Vec<(f64, f64)> = positions_two
+            .iter()
+            .zip(positions_one.iter())
+            .map(|((t1, d1), (_, d2))| (*t1, d1 - d2))
+            .collect();
+
+        self.plot_data(
+            "difference.png",
+            "Displacement Difference vs Time",
+            "Controller 0",
+            "Controller 1",
+            difference_one,
+            difference_two,
+        );
+    }
+
+    // Helper function to get the maximum displacement across both controllers
+    fn get_max_displacement(&self) -> f64 {
+        self.controller_one_pos
+            .iter()
+            .chain(self.controller_two_pos.iter())
+            .map(|(_, (x, y))| ((x.pow(2) + y.pow(2)) as f64).sqrt())
+            .fold(0.0, |a, b| a.max(b))
+    }
+}
+
+impl eframe::App for GameApp {
+    // Update method for the eframe application, called every frame
+    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
+        egui::CentralPanel::default().show(ctx, |ui| {
+            match &mut self.device_manager {
+                // device manager not yet initialized
+                None => match BackgroundMultiDevice::from_auto_configure(2) {
+                    Ok(device_manager) => {
+                        self.device_manager = Some(device_manager);
+                    }
+                    Err(e) => {
+                        display_welcome(ui, &e.to_string());
+                    }
+                },
```

Comment:
is there a reason this couldn't been done when initialization the eframe rather than having to run this check every update step?